lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,392 @@
+/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+/// [`Filter`]: lightning::chain::Filter
+pub struct EsploraSyncClient<L: Deref>
+where
+	L::Target: Logger,
+{
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
+	// Transactions that were previously processed, but must not be forgotten yet.
+	watched_transactions: Mutex<HashSet<Txid>>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	queued_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Outputs that were previously processed, but must not be forgotten yet.
+	watched_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Indicates whether we need to resync, e.g., after encountering an error.
+	pending_sync: AtomicBool,
+	// The tip hash observed during our last sync.
+	#[cfg(not(feature = "async-interface"))]
+	last_sync_hash: std::sync::Mutex<Option<BlockHash>>,
+	#[cfg(feature = "async-interface")]
+	last_sync_hash: futures::lock::Mutex<Option<BlockHash>>,
+	#[cfg(not(feature = "async-interface"))]
+	client: BlockingClient,
+	#[cfg(feature = "async-interface")]
+	client: AsyncClient,
+	logger: L,
+}
+
+impl<L: Deref> EsploraSyncClient<L>
+where
+	L::Target: Logger,
+{
+	/// Synchronizes the given confirmables via the [`Confirm`] interface. This method should be
+	/// called regularly to keep LDK up-to-date with current chain data.
+	///
+	/// [`Confirm`]: lightning::chain::Confirm
+	#[maybe_async]
+	pub fn sync(&self, confirmables: Vec<&(dyn Confirm + Sync + Send)>) -> Result<(), TxSyncError> {
+		log_info!(self.logger, "Starting transaction sync.");
+		// This lock makes sure we're syncing once at a time.
+		#[cfg(not(feature = "async-interface"))]
+		let mut locked_last_sync_hash = self.last_sync_hash.lock().unwrap();
+		#[cfg(feature = "async-interface")]
+		let mut locked_last_sync_hash = self.last_sync_hash.lock().await;
+
+		let mut tip_hash = maybe_await!(self.client.get_tip_hash())?;
+
+		loop {
+			let pending_sync = self.pending_sync.load(Ordering::SeqCst);
+			let pending_registrations = self.process_queues();
+			let tip_is_new = Some(tip_hash) != *locked_last_sync_hash;
+
+			// We loop until any registered transactions have been processed at least once, or the
+			// tip hasn't been updated during the last iteration.
+			if !pending_sync && !pending_registrations && !tip_is_new {
+				// Nothing to do.
+				break;
+			} else {
+				// Update the known tip to the newest one.
+				if tip_is_new {
+					// First check for any unconfirmed transactions and act on it immediately.
+					match maybe_await!(self.sync_unconfirmed_transactions(&confirmables)) {
+						Ok(()) => {},
+						Err(err) => {
+							// (Semi-)permanent failure, retry later.
+							log_error!(self.logger, "Failed during transaction sync, aborting.");
+							self.pending_sync.store(true, Ordering::SeqCst);
+							return Err(err);
+						}
```

Comment:
Yes, because it simply can't arise from `sync_unconfirmed_transactions`: the requests can either succeed or fail (=> `Failed`), but we get no information about any chain reorgs that happened during our sync run.